// @ledgerproof/sdk/dist/chunk-ZURYSARN.js
import { sha256 } from '@noble/hashes/sha256';
import { bytesToHex } from '@noble/hashes/utils';

// src/canonical.ts
function canonicalJson(obj) {
  return stringify(obj);
}
function stringify(value) {
  if (value === null || value === void 0) return "null";
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error("canonical JSON does not support NaN or Infinity");
    }
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    return escapeString(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(stringify).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value;
    const keys = Object.keys(obj).sort();
    return "{" + keys.filter((k) => obj[k] !== void 0).map((k) => escapeString(k) + ":" + stringify(obj[k])).join(",") + "}";
  }
  throw new Error(`cannot canonicalize value of type ${typeof value}`);
}
function escapeString(s) {
  const stdEscaped = JSON.stringify(s);
  let out = '"';
  for (let i = 1; i < stdEscaped.length - 1; i++) {
    const ch = stdEscaped[i];
    const code = ch.charCodeAt(0);
    if (code >= 32 && code <= 126) {
      out += ch;
    } else if (code >= 55296 && code <= 56319) {
      const high = code;
      const low = stdEscaped.charCodeAt(i + 1);
      out += "\\u" + high.toString(16).padStart(4, "0");
      out += "\\u" + low.toString(16).padStart(4, "0");
      i++;
    } else {
      out += "\\u" + code.toString(16).padStart(4, "0");
    }
  }
  out += '"';
  return out;
}
function sha256Hex(data) {
  const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data;
  return bytesToHex(sha256(bytes));
}
function contentHash(content) {
  return sha256Hex(canonicalJson(content));
}
function entryHash(entry) {
  return sha256Hex(canonicalJson(entry));
}
function artifactHash(artifact) {
  return sha256Hex(artifact);
}

export { artifactHash, canonicalJson, contentHash, entryHash, sha256Hex };
