// @ledgerproof/sdk/dist/chunk-VB27OW3R.js
import { artifactHash, canonicalJson, sha256Hex } from './chunk-ZURYSARN.js';
import * as ed from '@noble/ed25519';
import { sha512 } from '@noble/hashes/sha512';
import { hexToBytes, bytesToHex } from '@noble/hashes/utils';

// src/errors.ts
var LedgerProofError = class extends Error {
  context;
  constructor(message, context = {}) {
    super(message);
    this.name = "LedgerProofError";
    this.context = context;
    Object.setPrototypeOf(this, new.target.prototype);
  }
};
var ConfigurationError = class extends LedgerProofError {
  name = "ConfigurationError";
};
var AuthenticationError = class extends LedgerProofError {
  name = "AuthenticationError";
};
var RateLimitError = class extends LedgerProofError {
  name = "RateLimitError";
  retryAfterSeconds;
  constructor(message, options = {}) {
    super(message, options.context);
    this.retryAfterSeconds = options.retryAfterSeconds;
  }
};
var APIError = class extends LedgerProofError {
  name = "APIError";
  statusCode;
  body;
  requestId;
  constructor(message, options) {
    super(message, options.context);
    this.statusCode = options.statusCode;
    this.body = options.body;
    this.requestId = options.requestId;
  }
};
var NetworkError = class extends LedgerProofError {
  name = "NetworkError";
};
var ValidationError = class extends LedgerProofError {
  name = "ValidationError";
  errors;
  constructor(message, options = {}) {
    super(message, options.context);
    this.errors = options.errors ?? [];
  }
};
var GDPRSafetyError = class extends LedgerProofError {
  name = "GDPRSafetyError";
};
var KeyManagementError = class extends LedgerProofError {
  name = "KeyManagementError";
};
var ChainError = class extends LedgerProofError {
  name = "ChainError";
};
ed.etc.sha512Sync = (...m) => sha512(ed.etc.concatBytes(...m));
var Keypair = class _Keypair {
  constructor(privateKey) {
    this.privateKey = privateKey;
    if (privateKey.length !== 32) {
      throw new KeyManagementError("private key must be exactly 32 bytes", {
        gotLength: privateKey.length
      });
    }
  }
  privateKey;
  /** Generate a fresh random Ed25519 keypair via the platform CSPRNG. */
  static generate() {
    return new _Keypair(ed.utils.randomPrivateKey());
  }
  /** Construct from a 64-character hex-encoded 32-byte private key seed. */
  static fromHex(hex) {
    if (hex.length !== 64) {
      throw new KeyManagementError("signing key hex must be 64 characters", {
        gotLength: hex.length
      });
    }
    try {
      return new _Keypair(hexToBytes(hex));
    } catch (exc) {
      throw new KeyManagementError(`invalid hex: ${exc.message}`);
    }
  }
  /** Read `LEDGERPROOF_SIGNING_KEY_HEX` from the environment, or return null. */
  static fromEnv(varName = "LEDGERPROOF_SIGNING_KEY_HEX") {
    const value = typeof process !== "undefined" && process.env ? process.env[varName] : void 0;
    if (!value) return null;
    return _Keypair.fromHex(value.trim());
  }
  /** Sign `data` with this keypair, returning the raw 64-byte signature. */
  sign(data) {
    return ed.sign(data, this.privateKey);
  }
  /** Verify a signature over `data` using this keypair's public key. */
  verify(signature, data) {
    try {
      return ed.verify(signature, data, this.publicKey());
    } catch {
      return false;
    }
  }
  /** Derive the public key bytes (32 bytes). */
  publicKey() {
    return ed.getPublicKey(this.privateKey);
  }
  /** Base64-encoded public key, matching `POST /v1/keys` body format. */
  publicKeyBase64() {
    const pub = this.publicKey();
    if (typeof Buffer !== "undefined") {
      return Buffer.from(pub).toString("base64");
    }
    let binary = "";
    for (const b of pub) binary += String.fromCharCode(b);
    return btoa(binary);
  }
  /** Hex-encoded private key. Handle with care — never log or persist insecurely. */
  privateKeyHex() {
    return bytesToHex(this.privateKey);
  }
  toString() {
    return `Keypair(pub_b64=${this.publicKeyBase64().slice(0, 11)}\u2026)`;
  }
};

// src/transport.ts
var DEFAULT_MAX_RETRIES = 3;
var DEFAULT_TIMEOUT_MS = 3e4;
var RETRYABLE_STATUS = /* @__PURE__ */ new Set([429, 502, 503, 504]);
var Transport = class {
  apiBase;
  apiKey;
  publisherId;
  maxRetries;
  timeoutMs;
  userAgent;
  fetchImpl;
  constructor(options) {
    this.apiBase = options.apiBase.replace(/\/$/, "");
    this.apiKey = options.apiKey;
    this.publisherId = options.publisherId;
    this.maxRetries = options.maxRetries ?? DEFAULT_MAX_RETRIES;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.userAgent = options.userAgent ?? "ledgerproof-typescript/1.0.0 (web-fetch)";
    this.fetchImpl = options.fetch ?? globalThis.fetch;
    if (!this.fetchImpl) {
      throw new ConfigurationError(
        "no fetch implementation available; pass options.fetch or upgrade to Node 18+"
      );
    }
  }
  async request(opts) {
    const url = `${this.apiBase}${opts.path}`;
    const headers = {
      "User-Agent": this.userAgent,
      "Content-Type": "application/json",
      ...opts.headers ?? {}
    };
    if (opts.authenticated !== false) {
      if (!this.apiKey || !this.publisherId) {
        throw new ConfigurationError(
          "API key and publisherId required for authenticated requests"
        );
      }
      headers["X-Api-Key"] = this.apiKey;
      headers["X-Publisher-Id"] = this.publisherId;
    }
    const body = opts.json !== void 0 ? JSON.stringify(opts.json) : void 0;
    let lastError;
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeoutMs);
      let response;
      try {
        response = await this.fetchImpl(url, {
          method: opts.method,
          headers,
          body,
          signal: controller.signal
        });
      } catch (exc) {
        clearTimeout(timeoutId);
        lastError = exc;
        if (attempt < this.maxRetries) {
          await sleep(backoffDelay(attempt, null));
          continue;
        }
        throw new NetworkError(`network error contacting ${url}: ${lastError.message}`, {
          url
        });
      }
      clearTimeout(timeoutId);
      if (response.ok) {
        if (response.status === 204) return void 0;
        return await response.json();
      }
      if (RETRYABLE_STATUS.has(response.status) && attempt < this.maxRetries) {
        const retryAfter = response.headers.get("retry-after");
        await sleep(backoffDelay(attempt, retryAfter));
        continue;
      }
      await raiseForResponse(response, url);
    }
    throw new NetworkError(`exhausted retries to ${url}`, {
      cause: lastError?.message
    });
  }
};
async function raiseForResponse(response, url) {
  const requestId = response.headers.get("x-request-id") ?? void 0;
  const bodyText = await response.text().catch(() => "");
  let message = bodyText;
  let errors = [];
  try {
    const parsed = JSON.parse(bodyText);
    if (parsed?.error?.message) message = parsed.error.message;
    if (Array.isArray(parsed?.errors)) errors = parsed.errors;
  } catch {
  }
  const ctx = { status: response.status, requestId, url };
  if (response.status === 401 || response.status === 403) {
    throw new AuthenticationError(message, ctx);
  }
  if (response.status === 429) {
    const retryAfter = response.headers.get("retry-after");
    const retryAfterSeconds = retryAfter ? Number(retryAfter) : void 0;
    throw new RateLimitError(message, {
      retryAfterSeconds: Number.isFinite(retryAfterSeconds) ? retryAfterSeconds : void 0,
      context: ctx
    });
  }
  if (response.status === 400 || response.status === 422) {
    throw new ValidationError(message, { errors, context: ctx });
  }
  throw new APIError(message, {
    statusCode: response.status,
    body: bodyText,
    requestId,
    context: ctx
  });
}
function backoffDelay(attempt, retryAfter) {
  if (retryAfter) {
    const sec = Number(retryAfter);
    if (Number.isFinite(sec)) return Math.min(sec * 1e3, 3e4);
  }
  return Math.min(250 * 2 ** attempt, 3e4);
}
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// src/client.ts
var DEFAULT_API_BASE = "https://api-eu.ledgerproofhq.io";
var GENESIS_PREV_HASH = "0".repeat(64);
var TIP_PROBE_LIMIT = 1e4;
var EMAIL_REGEX = /[\w.+-]+@[\w-]+\.[\w.-]+/;
var LedgerProof = class {
  publisherId;
  deployerCountry;
  apiKey;
  apiBase;
  keyId;
  keypair;
  transport;
  keyRegistered = false;
  constructor(config) {
    if (config.publisherId.includes("@")) {
      throw new GDPRSafetyError(
        "publisher_id must be a legal-entity identifier (LEI/EUID/VAT/DID), not an email"
      );
    }
    this.publisherId = config.publisherId;
    this.deployerCountry = config.deployerCountry.toUpperCase();
    const apiKey = config.apiKey ?? (typeof process !== "undefined" ? process.env?.LEDGERPROOF_API_KEY : void 0);
    if (!apiKey) {
      throw new ConfigurationError(
        "API key required: pass config.apiKey or set LEDGERPROOF_API_KEY"
      );
    }
    this.apiKey = apiKey;
    this.apiBase = config.apiBase ?? (typeof process !== "undefined" ? process.env?.LEDGERPROOF_API_BASE : void 0) ?? DEFAULT_API_BASE;
    this.keyId = config.keyId ?? (typeof process !== "undefined" ? process.env?.LEDGERPROOF_KEY_ID : void 0) ?? "default";
    const signingHex = config.signingKeyHex ?? (typeof process !== "undefined" ? process.env?.LEDGERPROOF_SIGNING_KEY_HEX : void 0);
    this.keypair = signingHex ? Keypair.fromHex(signingHex) : Keypair.generate();
    this.transport = new Transport({
      apiBase: this.apiBase,
      apiKey: this.apiKey,
      publisherId: this.publisherId
    });
  }
  /** Issue an `ai/article-50/v1` receipt. The artifact is hashed locally. */
  async publishAiArticle50(opts) {
    if (EMAIL_REGEX.test(opts.deployerName)) {
      throw new GDPRSafetyError("deployerName looks like an email \u2014 use a legal name");
    }
    const artifactBytes = typeof opts.artifact === "string" ? new TextEncoder().encode(opts.artifact) : opts.artifact;
    const aHash = artifactHash(artifactBytes);
    const content = {
      ai_system_id: opts.aiSystemId,
      ai_system_version: opts.aiSystemVersion,
      deployer_id: this.publisherId,
      deployer_name: opts.deployerName,
      deployer_country: this.deployerCountry,
      content_category: opts.contentCategory,
      artifact_hash: aHash,
      artifact_content_type: opts.artifactContentType,
      artifact_bytes: artifactBytes.length,
      supervisory_authority: opts.supervisoryAuthority,
      generation_type: opts.generationType,
      source_content_hash: opts.sourceContentHash,
      perceptual_hash: opts.perceptualHash,
      transparency_marker: opts.transparencyMarker ?? "LPR-EU-AI-ACT-50",
      is_public_interest: opts.isPublicInterest,
      enforcement_date: opts.enforcementDate ?? "2026-08-02",
      profile_version: opts.profileVersion ?? "EU-AI-ACT-50-v1.1"
    };
    const cleaned = Object.fromEntries(
      Object.entries(content).filter(([, v]) => v !== void 0 && v !== null)
    );
    return this.publish("ai/article-50/v1", cleaned);
  }
  /** Fetch an entry from the public verifier endpoint. No auth required. */
  async verify(sequence) {
    return this.transport.request({
      method: "GET",
      path: `/v1/entries/${sequence}`,
      authenticated: false
    });
  }
  /** Look up receipts by SHA-256 of artifact content. Unauthenticated. */
  async lookupByContentHash(contentHash) {
    const data = await this.transport.request({
      method: "GET",
      path: `/v1/receipts/by-content-hash/${contentHash}`,
      authenticated: false
    });
    return data?.matches ?? [];
  }
  async publish(contentType, contentDict) {
    await this.ensureKeyRegistered();
    let lastError;
    for (let attempt = 0; attempt < 3; attempt++) {
      const { sequence, prevHash } = await this.discoverChainTip();
      const payload = this.buildPublishPayload(contentDict, contentType, sequence, prevHash);
      try {
        const response = await this.transport.request({
          method: "POST",
          path: "/v1/publish",
          json: payload
        });
        return {
          sequence: response.sequence,
          entry_hash: response.entry_hash,
          receipt_id: response.receipt_id,
          verify_url: `${this.apiBase}/v1/verify/${response.sequence}`
        };
      } catch (exc) {
        lastError = exc;
        if (exc instanceof ValidationError && attempt < 2 && (exc.message || "").toLowerCase().includes("sequence")) {
          await new Promise((r) => setTimeout(r, 200 * (attempt + 1)));
          continue;
        }
        throw exc;
      }
    }
    throw new ChainError(`could not publish after 3 attempts: ${lastError?.message}`);
  }
  buildPublishPayload(contentDict, contentType, sequence, prevHash) {
    const contentCanonical = canonicalJson(contentDict);
    const cHash = sha256Hex(contentCanonical);
    const entryTimestamp = isoNowMillis();
    const entryObj = {
      content: contentDict,
      content_hash: cHash,
      content_type: contentType,
      entry_timestamp: entryTimestamp,
      key_id: this.keyId,
      prev_hash: prevHash,
      protocol_version: "ledgerproof/1.0",
      publisher_id: this.publisherId,
      sequence
    };
    const entryCanonical = canonicalJson(entryObj);
    const eHash = sha256Hex(entryCanonical);
    const entryHashBytes = hexToBytes2(eHash);
    const signature = this.keypair.sign(entryHashBytes);
    const signatureHex = bytesToHex2(signature);
    return {
      publisher_id: this.publisherId,
      key_id: this.keyId,
      prev_hash: prevHash,
      entry_hash: eHash,
      signature: signatureHex,
      protocol_version: "ledgerproof/1.0",
      content_type: contentType,
      content_hash: cHash,
      content: contentDict,
      entry_json_canonical: entryCanonical,
      entry_timestamp: entryTimestamp
    };
  }
  async ensureKeyRegistered() {
    if (this.keyRegistered) return;
    try {
      await this.transport.request({
        method: "POST",
        path: "/v1/keys",
        json: {
          key_id: this.keyId,
          verifying_key_b64: this.keypair.publicKeyBase64()
        }
      });
    } catch (exc) {
      if (!(exc instanceof LedgerProofError)) throw exc;
    }
    this.keyRegistered = true;
  }
  async discoverChainTip() {
    let probe = 0;
    while (probe <= TIP_PROBE_LIMIT) {
      try {
        await this.transport.request({
          method: "GET",
          path: `/v1/entries/${probe}`,
          authenticated: false
        });
        probe++;
      } catch (exc) {
        const status = exc instanceof Object && "statusCode" in exc ? exc.statusCode : void 0;
        if (status === 404) {
          if (probe === 0) return { sequence: 0, prevHash: GENESIS_PREV_HASH };
          const prev = await this.transport.request({
            method: "GET",
            path: `/v1/entries/${probe - 1}`,
            authenticated: false
          });
          return { sequence: probe, prevHash: prev.entry_hash };
        }
        throw exc;
      }
    }
    throw new ChainError(`chain tip not found within ${TIP_PROBE_LIMIT} probes`);
  }
};
function isoNowMillis() {
  const d = /* @__PURE__ */ new Date();
  return d.toISOString().replace(/\.\d{3}Z$/, (m) => m);
}
function hexToBytes2(hex) {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}
function bytesToHex2(bytes) {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}

export { APIError, AuthenticationError, ChainError, ConfigurationError, DEFAULT_API_BASE, GDPRSafetyError, KeyManagementError, Keypair, LedgerProof, LedgerProofError, NetworkError, RateLimitError, Transport, ValidationError };
