// @ledgerproof/sdk/dist/verifier.cjs
'use strict';

var sha256 = require('@noble/hashes/sha256');
var utils = require('@noble/hashes/utils');

// src/canonical.ts
function sha256Hex(data) {
  const bytes = typeof data === "string" ? new TextEncoder().encode(data) : data;
  return utils.bytesToHex(sha256.sha256(bytes));
}

// src/verifier.ts
var DEFAULT_API_BASE = "https://api-eu.ledgerproofhq.io";
async function verifyReceipt(sequence, options = {}) {
  const apiBase = (options.apiBase ?? DEFAULT_API_BASE).replace(/\/$/, "");
  const fetchImpl = options.fetch ?? globalThis.fetch;
  const response = await fetchImpl(`${apiBase}/v1/entries/${sequence}`);
  if (response.status === 404) return null;
  if (!response.ok) {
    throw new Error(`verify failed: ${response.status} ${response.statusText}`);
  }
  return await response.json();
}
async function lookupByContentHash(contentHash, options = {}) {
  const apiBase = (options.apiBase ?? DEFAULT_API_BASE).replace(/\/$/, "");
  const fetchImpl = options.fetch ?? globalThis.fetch;
  const response = await fetchImpl(`${apiBase}/v1/receipts/by-content-hash/${contentHash}`);
  if (response.status === 404) return [];
  if (!response.ok) {
    throw new Error(`lookup failed: ${response.status}`);
  }
  const data = await response.json();
  return data?.matches ?? [];
}
async function hashArtifact(artifact) {
  let bytes;
  if (typeof artifact === "string") {
    bytes = new TextEncoder().encode(artifact);
  } else if (artifact instanceof Blob) {
    bytes = new Uint8Array(await artifact.arrayBuffer());
  } else if (artifact instanceof ArrayBuffer) {
    bytes = new Uint8Array(artifact);
  } else {
    bytes = artifact;
  }
  if (globalThis.crypto?.subtle?.digest) {
    const buf = bytes.buffer instanceof ArrayBuffer ? bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) : new Uint8Array(bytes).buffer;
    const digest = await globalThis.crypto.subtle.digest("SHA-256", buf);
    return Array.from(new Uint8Array(digest)).map((b) => b.toString(16).padStart(2, "0")).join("");
  }
  return sha256Hex(bytes);
}
async function scanPageForReceipts(options = {}) {
  if (typeof document === "undefined") return [];
  const elements = Array.from(document.querySelectorAll("[data-lpr-receipt]"));
  return Promise.all(
    elements.map(async (element) => {
      const seqAttr = element.getAttribute("data-lpr-receipt");
      const sequence = seqAttr ? parseInt(seqAttr, 10) : Number.NaN;
      if (!Number.isFinite(sequence)) {
        return { element, entry: null };
      }
      const entry = await verifyReceipt(sequence, options).catch(() => null);
      return { element, entry };
    })
  );
}

exports.DEFAULT_API_BASE = DEFAULT_API_BASE;
exports.hashArtifact = hashArtifact;
exports.lookupByContentHash = lookupByContentHash;
exports.scanPageForReceipts = scanPageForReceipts;
exports.verifyReceipt = verifyReceipt;
