// @ledgerproof/sdk/dist/adapters/openai.js
import { LedgerProofError, LedgerProof } from '../chunk-VB27OW3R.js';
import '../chunk-ZURYSARN.js';

// src/adapters/openai.ts
var PATCHED = /* @__PURE__ */ new WeakMap();
function attach(client, options) {
  if (!client?.chat?.completions?.create) {
    throw new LedgerProofError(
      "target does not look like an OpenAI client; pass an openai.OpenAI() instance"
    );
  }
  if (PATCHED.has(client)) return;
  const lp = new LedgerProof({
    publisherId: options.publisherId,
    deployerCountry: options.deployerCountry,
    ...options.apiKey !== void 0 && { apiKey: options.apiKey },
    ...options.apiBase !== void 0 && { apiBase: options.apiBase }
  });
  const original = client.chat.completions.create.bind(client.chat.completions);
  const wrapped = async (...args) => {
    const result = await original(...args);
    const callArgs = args[0] ?? {};
    const isStream = Boolean(callArgs.stream);
    if (isStream) {
      return wrapAsyncStream(result, lp, options, callArgs);
    }
    const text = extractText(result);
    if (text) {
      attachReceiptPromise(result, lp, options, callArgs, text);
    }
    return result;
  };
  client.chat.completions.create = wrapped;
  PATCHED.set(client, { original, completions: client.chat.completions });
}
function detach(client) {
  const entry = PATCHED.get(client);
  if (!entry) return;
  entry.completions.create = entry.original;
  PATCHED.delete(client);
}
function attachReceiptPromise(response, lp, options, callArgs, text) {
  const promise = issueSafe(lp, options, callArgs, text);
  try {
    response._ledgerproofPromise = promise;
  } catch {
  }
}
async function issueSafe(lp, options, callArgs, text) {
  try {
    const model = typeof callArgs.model === "string" ? callArgs.model : "unknown";
    const aiSystemId = options.aiSystemId ?? `openai/${model}`;
    return await lp.publishAiArticle50({
      artifact: text,
      artifactContentType: "text/plain",
      aiSystemId,
      deployerName: options.deployerName,
      contentCategory: "SYNTHETIC_TEXT",
      generationType: "FULLY_GENERATED",
      ...options.isPublicInterest !== void 0 && { isPublicInterest: options.isPublicInterest }
    });
  } catch {
    return null;
  }
}
function extractText(response) {
  try {
    const choice = response?.choices?.[0];
    const content = choice?.message?.content;
    if (typeof content === "string") return content;
    if (Array.isArray(content)) {
      return content.filter((p) => p?.type === "text").map((p) => p.text ?? "").join("");
    }
    return "";
  } catch {
    return "";
  }
}
function extractTextFromChunk(chunk) {
  try {
    return chunk?.choices?.[0]?.delta?.content ?? "";
  } catch {
    return "";
  }
}
function wrapAsyncStream(stream, lp, options, callArgs) {
  const chunks = [];
  const originalIterator = stream[Symbol.asyncIterator].bind(stream);
  return new Proxy(stream, {
    get(target, prop) {
      if (prop === Symbol.asyncIterator) {
        return () => ({
          async next() {
            const it = originalIterator();
            const result = await it.next();
            if (result.done) {
              const full = chunks.join("");
              if (full) {
                attachReceiptPromise(stream, lp, options, callArgs, full);
              }
              return result;
            }
            const text = extractTextFromChunk(result.value);
            if (text) chunks.push(text);
            return result;
          }
        });
      }
      return target[prop];
    }
  });
}

export { attach, detach };
